### ISSUE TITLE
Add support for jti claim

### ISSUE THREAD

**Issue by avimeir** [GetBlimp/django-rest-framework-jwt#89]
Hi all,

I'd like to add the support for jti claim to the package. The idea basically would be to store a list of expired jti's somewhere on the server side, and allow to blacklist specific tokens.

The way I'm thinking about the implementation is:

- Use pymongo to store the blacklist (rather than a django specific mongo package, makes our solution portable to vanilla python or other frameworks)
- Generate a long enough jti claim that will have negligible probability of being produced twice before the first one expired (string of 20 random ascii_letters and digits maybe?)
- Store the jti and the entire payload in mongodb, this will help us clean up the collection when a JWT is past the expiry date (so we don't grow the collection ad infinitum)
- Add a flag in api_settings.JWT_ENABLE_BLACKLIST, if True we add the jti claim to the payload in jwt_payload_handler()
- Check against the mongodb collection everytime we decode a JWT (if JWT_ENABLE_BLACKLIST is True) and if the jti is blacklisted treat it as if the JWT is expired

I have made an initial commit to a fork of the package, with my suggestion on how to implement this, you can find it here: https://github.com/avimeir/django-rest-framework-jwt/commits/master

Happy to discuss this!

Cheers,
Avi

**Comment by jpadilla** [comment]
@erichonkanen maybe you'd like to check this out

**Comment by erichonkanen** [comment]
Interesting, I hadn't heard of [JWT ID](http://tools.ietf.org/html/draft-ietf-oauth-json-web-token-32#section-4.1.7) before...  seems like this overlaps with my earlier search for a way to invalidate a token. What are your thoughts on adding mongo as a dependency to this package? The implementation looks good at first glance

**Comment by jpadilla** [comment]
Mongo as a dependency is definitely a no. We should probably just include a default implementation that uses Django cache backend, like how [throttling](http://www.django-rest-framework.org/api-guide/throttling/) in DRF works.

**Comment by gcollazo** [comment]
I believe Django cache is not the right solution since cache contents can be evicted or lost at any time. If I understand correctly, when you blacklist a token that token should be blacklisted forever. In that case using the DB is right choice.

I suggest creating a model `DRFJWTBlacklist` and store blacklisted tokens there. With this solution enabling a token blacklist feature will require DB access on all protected HTTP request. This should be an option and not the default behavior.

**Comment by erichonkanen** [comment]
@gcollazo what are you suggesting for the default behavior?

**Comment by jpadilla** [comment]
Yeah that sounds about right. But having said that we should provide handlers like what @avimeir did. We shouldn't care where the user stores and checks for blacklisted tokens. Optional default implementation could definitely be provide a model, properly indexed, unique constraints, etc.

Also, I think UUID's might better for the actual `jti`.

**Comment by gcollazo** [comment]
:+1: for what @jpadilla said. Consumers of this package should be able to store and check for backlisted token wherever they want.

Providing a basic implementation that used django models sounds like a good idea but should be disabled by default. Users must enable this feature (if they want it) and specify if they want to use the bundled model backed implementation of token blacklist or use a custom one they can implement themselves.

**Comment by jpadilla** [comment]
Good read about [Blacklisting JSON Web Tokens](https://auth0.com/blog/2015/03/10/blacklist-json-web-token-api-keys/).

**Comment by avimeir** [comment]
So to summarize the main changes you would make:
1. move from mongo to a model that uses the default django ORM settings
2. use UUID for the `jti` instead of the random string I've created

Am I missing something?

Happy to fix my code and issue a pull request, unless someone from the core contributors wants to take ownership. As long as it gets done and we can use it in our project - all solutions are fine with me

**Comment by jpadilla** [comment]
@avimeir yeah that sounds good. Feel free to submit a PR. No one else is working on this right now.

**Comment by erichonkanen** [comment]
I pulled down @avimeir code so far and tweaking it to use a model with UUIDField (either through django-uuidfield or native if > 1.8).. seeing if I can get it working etc.. if @avimeir has update before then i'll check it out

**Comment by erichonkanen** [comment]
@avimeir @jpadilla touched up the code to provide default get and set handlers for blacklisting and a default model implementation... let me know your thoughts

https://github.com/GetBlimp/django-rest-framework-jwt/commit/0c0f946d8ff72c2ac4b1c0425649b14a29f493b6

PS havent tested yet but will this weekend

**Comment by tachang** [comment]
Is JTI meant to be used as a session identifier?

**Comment by erichonkanen** [comment]
@tachang per the spec:

```
The jti (JWT ID) claim provides a unique identifier for the JWT. The identifier value MUST be assigned in a manner that ensures that there is a negligible probability that the same value will be accidentally assigned to a different data object; if the application uses multiple issuers, collisions MUST be prevented among values produced by different issuers as well.
```

So essentially yes, which might be confusing since the whole point of using jwt is for a stateless auth mechanism... [This](https://auth0.com/blog/2015/03/10/blacklist-json-web-token-api-keys/) page helped...

**Comment by tachang** [comment]
Cool. Makes a lot of sense. Kind of like old school session cookies where you stored each cookie in the database. Or a certificate revocation list.

I like the idea of using Django's default database since the biggest issue with blacklists is always propagating them.

Just a thought but the main reason I'd want to invalidate a token is because a user is marked inactive. I assume then I'd have to iterate through all tokens generated for a specific user and then add them to the blacklist? This is probably the most common use case for this feature.

**Comment by erichonkanen** [comment]
@tachang yeah I think you're right about the primary reason for invalidating a token is an inactive user... previously I was thinking of implementing an activity-based auth mechanism that could log out a user and invalidate the jwt and this seems precisely fit for that...

If you are storing all the tokens generated in the db then yes you might want to iterate through any previous non-expired ones and blacklist them. However at least in my case I don't store the tokens at all and would probably just use the default implementation here where a blacklist token obj is created manually when blacklisting a token.

**Comment by avimeir** [comment]
@jpadilla  Any plan to merge this into a future release?

**Comment by erichonkanen** [comment]
@avimeir sorry this is taking so long but I think we'll get it in soon.. pending hooking it into a demo app

**Comment by avimeir** [comment]
Thanks!

**Comment by jpadilla** [comment]
After taking a look at the awesome demo put together by @erichonkanen and talking with him at the different use cases for this feature and how it fits together with this package I'm inclining more that this feature be released as a separate package. I'm failing to see this now fit into the purpose of this(django-rest-framework-jwt) package which is basically just enabling auth with JWT.

There are many benefits of this being released as another package. We don't currently have a use case for the feature so maintaining further along it would be harder. @avimeir Would you be interested in taking charge? All of the work is already there, it just needs to be placed in a new package and someone willing to maintain it.

**Comment by erichonkanen** [comment]
additionally, i'd like to hear more about the use cases and thoughts on validity of doing jwt revocation after reading conflicting accounts online...

**Comment by avimeir** [comment]
@erichonkanen's demo is really cool, and I'd keep it anyway, maybe evolving it into a demo explaining JWT for new users?

As per blacklisting- my main use case was to force logout users even if they hold a valid JWT. I will consider taking it out to a new package, especially if more people are interested in it (currently I'm using a fork that enables this, but it's not open-source ready yet). 

Anyway, I agree that it might fit better outside of django-rest-framework-jwt. Thank guys!

**Comment by erichonkanen** [comment]
@avimeir Im curious, @jpadilla brought up the fact that an alternate approach to force-logging them out would be to just delete their jwt from the client storage/cache, what are your thoughts on that?

**Comment by avimeir** [comment]
Won't work for applications where you must ensure that the user is logged out indeed, a smart user could copy the token client-side and keep using it until it expires.

**Comment by erichonkanen** [comment]
that makes sense... so yeah I think the blacklist feature makes sense still but haven't heard anyone else wanting to use it which makes me wonder if it would make sense in the main repo.. I do like the fact that it is just optional anyways so it is opt-in

if anyone else watching and has feedback plz reply

**Comment by gabn88** [comment]
Why not check if the user is active after the authentication is done? This seems an authorization question, rather then an authentication question. Just my 2 cents.

**Comment by erichonkanen** [comment]
@gabn88 this can probably be closed... I did end up implementing a blacklist token feature but we decided it wasn't best to be included in the main package, and also at the time I personally didn't have the need or understanding to use it (although it did work nice).

**Comment by dheavy** [comment]
Sorry to bump this old thread, but I still find the JWT blacklisting highly valuable. Best use case is a mitigation scenario whenever a token was compromised (and maybe you'd have the user force all connected devices/browser to log out before changing her credentials and logging in again from a safe place).
@erichonkanen I'm about to dig up your old PR and try to adapt your program to the changed codebase. Should I know anything before I do? (I'm thinking maybe you've learned new valuable things since then).

Cheers

**Comment by erichonkanen** [comment]
@dheavy go right ahead! one thing I did that might impact it is I think I abstracted part of the view stuff out from how it was before... i'd have to look into it again and Im not sure what has changed since then but let me know if you want me to look at anything! it was working well at time it was submitted...

**Comment by erichonkanen** [comment]
@dheavy https://github.com/erichonkanen/django-rest-framework-jwt/blob/4306ab9916a82e77c29c30256249b4e4f178c251/rest_framework_jwt/views.py#L59 (and L19) I had made those staticmethod so you can override it easily in a subclassed view such as [here](https://github.com/erichonkanen/django-rest-framework-jwt/blob/4306ab9916a82e77c29c30256249b4e4f178c251/rest_framework_jwt/blacklist/views.py#L14), that said it was just an idea at the time!

**Comment by angvp** [comment]
Can we close this one?
